src/github.com/couchbase/sync_gateway/base/bucket_gocb.go

```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
+	if result == nil {
+		return map[string][]byte{}, err
+	}
+
+	// Type assertion of result into a map
+	resultMap, ok := result.(map[string][]byte)
+	if !ok {
+		LogPanic("Error doing type assertion of %v into a map", result)
+	}
+
+	return resultMap, err
+
+}
+
+func (bucket CouchbaseBucketGoCB) newGetBulkRawRetryWorker(keys []string) RetryWorker {
+
+	// resultAccumulator scoped in closure, will accumulate results across multiple worker invocations
+	resultAccumulator := make(map[string][]byte, len(keys))
+
+	// pendingKeys scoped in closure, represents set of keys that still need to be attempted or re-attempted
+	pendingKeys := keys
+
+	worker := func() (shouldRetry bool, err error, value interface{}) {
+
+		retryKeys := []string{}
+		keyBatches := createBatchesKeys(MaxBulkBatchSize, pendingKeys)
+		for _, keyBatch := range keyBatches {
+
+			// process batch and add successful results to resultAccumulator
+			// and recoverable (non "Not Found") errors to retryKeys
+			err := bucket.processBatch(keyBatch, resultAccumulator, retryKeys)
+			if err != nil {
+				return false, err, nil
+			}
 
-		var items []gocb.BulkOp
-		for _, key := range batchKeys {
-			var value []byte
-			item := &gocb.GetOp{Key: key, Value: &value}
-			items = append(items, item)
 		}
-		err := bucket.Do(items)
-		if err != nil {
-			Warn("Bulk get error for batch size %d:%v", len(items), err)
-			return nil, err
+
+		// if there are no keys to retry, then we're done.
+		if len(retryKeys) == 0 {
+			return false, nil, resultAccumulator
 		}
 
-		for _, item := range items {
-			getOp, ok := item.(*gocb.GetOp)
-			if !ok {
-				continue
+		// otherwise, retry the keys the need to be retried
+		keys = retryKeys
+
+		// return true to signal that this function needs to be retried
+		return true, nil, nil
+
+	}
+
+	return worker
+
+}
+
+func (bucket CouchbaseBucketGoCB) processBatch(keys []string, resultAccumulator map[string][]byte, retryKeys []string) error {
+
+	var items []gocb.BulkOp
+	for _, key := range keys {
+		var value []byte
+		item := &gocb.GetOp{Key: key, Value: &value}
+		items = append(items, item)
+	}
+	err := bucket.Do(items)
+	if err != nil {
+		return err
+	}
+
+	for _, item := range items {
+		getOp, ok := item.(*gocb.GetOp)
+		if !ok {
+			continue
+		}
+		// Ignore any ops with errors.
+		// NOTE: some of the errors are misleading:
+		// https://issues.couchbase.com/browse/GOCBC-64
+		if getOp.Err == nil {
+			byteValue, ok := getOp.Value.(*[]byte)
+			if ok {
+				resultAccumulator[getOp.Key] = *byteValue
 			}
-			// Ignore any ops with errors.
-			// NOTE: some of the errors are misleading:
-			// https://issues.couchbase.com/browse/GOCBC-64
-			if getOp.Err == nil {
-				byteValue, ok := getOp.Value.(*[]byte)
-				if ok {
-					result[getOp.Key] = *byteValue
-				}
+		} else {
+			// if it's a recoverable error, then throw it in retry collection.
+			if isRecoverableGoCBError(getOp.Err) {
```

Comment:
Related to the question below - why isRecoverableGoCBError here, and not shouldRetryGoCBOp?